### PR TITLE
Complete PHP-MD to ECS migration

### LINKED ISSUE #7: Make use of slevomat and object-calisthenics to move from phpmd
These rules of `phpmd`:

- `ShortVariable`
- `ShortMethodName`
- `UnusedPrivateMethod`
- `ExcessivePublicCount`
- `ExcessiveClassLength`
- `ExcessiveMethodLength`

**are already implemented in these 2 packages** for PHP_CodeSniffer as Sniffs:

- https://github.com/object-calisthenics/phpcs-calisthenics-rules - you gonna love it, if you like strictness, and I can see you do :)
- https://github.com/slevomat/coding-standard

And few more useful rules, that could be added to `rulesetCS.xml`. Both packages are well maintained and up-to-date with PHP_CodeSniffer 3.x.

By using them, you could easily drop `phpmd` that never fix the code for you and move to `PHP_CodeSniffer` and `PHP CS Fixer` who do.

### PR DESCRIPTION
**Pull request by TomasVotruba**
**Description, reason for the PR:** Follow-up to #12, closes #7 

**New feature:** No

**BC breaks:** No

**Standards and tests pass:** Yes

**Licence:** MIT

### Related Changes

- Added `CamelCaseParameterSniff` that checks `$camelCase` naming for variables, see [failing Travis](https://travis-ci.org/shopsys/coding-standards/jobs/327019302#L621)
  
- Added [`/examples`](/examples) directory to demonstrate valid code for this standard
  

### Todo

- [x] make Sniff fixable or Fixer
- [x] retarged branch to `master` after #12 is merged

### CODE REVIEW

**Comment by TomasVotruba** [comment]
Ready to review and merge

File: .gitignore [hunk 1]
```diff
@@ -1,5 +1,2 @@
-/.idea
-/composer.lock
-/nbproject
```
Comment (TomasVotruba) [inline .gitignore]: IDE files should be ignored globally once on one's pc, not in every project for every IDE.

https://stackoverflow.com/a/7335487/1348344

File: src/CsFixer/OrmJoinColumnRequireNullableFixer.php [hunk 2]
```diff
@@ -24,16 +24,23 @@ public function getDefinition(): FixerDefinitionInterface
...
+SAMPLE
+            ), new CodeSample(
+<<<'SAMPLE'
+/**
+ * @var \StdObject
+ * @ORM\OneToOne(targetEntity="StdObject")
+ */
+private $foo;
+SAMPLE
+            )]
```
Comment (TomasVotruba) [inline src/CsFixer/OrmJoinColumnRequireNullableFixer.php]: This should be much more readable then inlined escaped string

File: src/Sniffs/CamelCaseParameterSniff.php [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    /**
+     * Changes:
+     * _someVariable => someVariable
+     * some_variable => someVariable
+     */
+    private function fixVariableName(File $file, int $position, string $variableName): void
+    {
+        $newVariableName = preg_replace_callback(
+            '#(?<prefixLetter>[a-z]{1})?' . self::NOT_LETTER_THEN_LETTER_PATTERN . '#',
+            function (array $match): string {
```
Comment (TomasVotruba) [inline src/Sniffs/CamelCaseParameterSniff.php]: Let me know if comment + method is clear in what it does. 
I'll improve the naming and add more examples if not

Comment (vitek-rostislav) [reply src/Sniffs/CamelCaseParameterSniff.php]: We discussed this with @PetrHeinz and he was not satisfied with the solution (eg. we came up with few use cases when this solution is not sufficient and it does not take numbers into account). We suggest to copypaste the `PHP_CodeSniffer\Standards\Squiz\Sniffs\NamingConventions\ValidVariableNameSniff` into new `ShopSys\CodingStandards\Sniffs\ValidVariableNameSniff` and just remove the logic of ignoring private fields prefixed with underscore. What do you think?

**Review by vitek-rostislav — CHANGES_REQUESTED** [review]

File: .gitignore (hunk 1 above)
Comment (TomasVotruba) [reply .gitignore]: I see

File: src/Sniffs/CamelCaseParameterSniff.php (hunk 3 above)
Comment (TomasVotruba) [reply src/Sniffs/CamelCaseParameterSniff.php]: I see. Could you share these use cases? That's the most important information here.

`PHP_CodeSniffer\Standards\Squiz\Sniffs\NamingConventions\ValidVariableNameSniff` is already used actually. **This sniff takes care about `_` prefixes  in private properties.** Together, they should cover it all. If not, it's a bug that I shall fix.

Copy-pasting whole sniff in from 3rd party is fast idea, that might fix it the issue today.  But eventually could lead to:

- missing update
- missing tests
- lot of code that is not understandable and difficult to maintain and develop, which PHP_CodeSniffer code sure is

We could use original Sniff thought, that's a good idea. There is advanced  regular that checks this: https://github.com/squizlabs/PHP_CodeSniffer/blob/818f188f71940f9cf848eca29b037cbaa2cde2f2/src/Standards/Squiz/Sniffs/NamingConventions/ValidVariableNameSniff.php#L188

But, if the reported cases are nearly impossible to fix, we might give it a try :+1:

Comment (TomasVotruba) [reply src/Sniffs/CamelCaseParameterSniff.php]: @vitek-rostislav Btw, this is why story is so important. 

- *It doesn't work => solution* is like "git squash to 1 commit"
- *It dones't work on these cases => this is probably the best solution* is like having the full commit hiStory :+1:

Comment (vitek-rostislav) [reply src/Sniffs/CamelCaseParameterSniff.php]: Ok, fair enough, you are right that copypasting 3rd party code is not a good solution.

One of the problems is that there is used method `Common::isCamelCaps($variableName)` to determine whether the variable is in camelCase and than there is used different logic for fixing the variable which does not take into account numbers, PHP superglobals (even though they are covered by different sniff), unicode symbols, and edge cases like `$$var` or having `$aB` and `$a_b` in the same file, where fixing of such thing could break the application. Maybe the main problem here is that we do not have specified all use cases that we want to cover with the fixer. So I suggest not to implement the fix at the time, and we can do it afterwards we define the use cases properly.

Unfortunately, I also found another problem with the sniff itself, it does not report declaration of private variable in underscore as a violation.

Comment (TomasVotruba) [reply src/Sniffs/CamelCaseParameterSniff.php]: > So I suggest not to implement the fix at the time, and we can do it afterwards we define the use cases properly.

All right, that makes sense. I'll remove it for now then

Comment (TomasVotruba) [reply src/Sniffs/CamelCaseParameterSniff.php]: Done :+1:

**Review by vitek-rostislav — APPROVED** [review]

**Comment by TomasVotruba** [comment]
I've also talked with CodeSniffer author about improving the variable sniff and created an issue on his recommendation: https://github.com/squizlabs/PHP_CodeSniffer/issues/1851

**Comment by TomasVotruba** [comment]
@vitek-rostislav Sure, go with it!
I'm not on PC now, but you can do it on merge

**Comment by TomasVotruba** [comment]
@vitek-rostislav Great job :+1:

**Pull request merged**